csharp/ql/src/semmle/code/csharp/AnnotatedType.qll

```diff
@@ -0,0 +1,278 @@
+/**
+ * Provides predicates and classes for representing annotations on types, such as nullability information.
+ *
+ * Use an element's `getType()` predicate to access the standard type of the element, without regard
+ * for annotations on the type. Use the corresponding `getAnnotatedType()` predicate to
+ * get the annotated type, that includes this additional information.
+ */
+
+import csharp
+
+module Annotations {
+  private newtype TAnnotation =
+    TDisabledNullability() or
+    TNotApplicableNullability() or
+    TNotNullableRefType() or
+    TNullableRefType() or
+    TReadonlyRefType() or
+    TRefType() or
+    TOutType()
+
+  /** An annotation on a type. */
+  class TypeAnnotation extends TAnnotation {
+    /** Holds if this string should prefix the type string. */
+    predicate isPrefix() { none() }
+
+    /** Holds if this string should suffix the type string. */
+    predicate isSuffix() { none() }
+
+    /** Gets the bit position in the bit-field. */
+    abstract int getBit();
```

Comment:
I suggest having a separate set of `Impl` classes for the internal predicates, i.e.
```
private class TypeAnnotationImpl extends TypeAnnotation {
    /** Gets the bit position in the bit-field. */
    abstract int getBit();
}

private class DisabledNullabiltyImpl extends TypeAnnotationImpl, DisabledNullabilty {
    override int getBit() { result = 0 }
}

...
```

Reply:
I have resolved this issue (and the `private` comments below) by making the entire `Annotations` module `private`. The user can just use `AnnotatedType.isRef()` etc, rather than to expose the annotation classes. Less is more.